src/crypto/tls/ech.go

```diff
@@ -0,0 +1,874 @@
+	}
+
+	if innerDigest != nil {
+		if subtle.ConstantTimeCompare(innerDigest, ctx.innerDigest(rawHelloInner)) != 1 {
+			c.sendAlert(alertDecryptError)
+			return nil, fmt.Errorf("ech: outer extensions: inner digest validation fails")
+		}
+	}
+
+	helloInner := new(clientHelloMsg)
+	if !helloInner.unmarshal(rawHelloInner) {
+		c.sendAlert(alertIllegalParameter)
+		return nil, fmt.Errorf("ech: %s", err)
+	}
+
+	// Update the HRR pre-shared-key. This is used to decrypt the second
+	// ClientHelloInner in case an HRR is sent.
+	c.ech.hrrPsk = ctx.hrrPsk()
+
+	// Accept ECH.
+	c.ech.st.Accepted = true
+	return helloInner, nil
+}
+
+// echCipherSuite represents an ECH ciphersuite, a KDF/AEAD algorithm pair. This
+// is different from an HPKE ciphersuite, which represents a KEM, KDF, and an
+// AEAD algorithm.
+type echCipherSuite struct {
+	kdfId, aeadId uint16
+}
+
+// echAssembleHpkeCipherSuite maps the codepoints for an HPKE ciphersuite to the
+// ciphersuite's internal representation, verifying that the host supports the
+// cipher suite.
+func echAssembleHpkeCipherSuite(kemId, kdfId, aeadId uint16) (hpke.CipherSuite, error) {
+	// Verify that the ciphersuite is supported by github.com/cisco/go-hpke and
+	// return the ciphersuite's internal representation.
+	return hpke.AssembleCipherSuite(hpke.KEMID(kemId), hpke.KDFID(kdfId), hpke.AEADID(aeadId))
+}
+
+// echUnmarshalHpkePublicKey parses a serialized public key for the KEM algorithm
+// identified by `kemId`.
+func echUnmarshalHpkePublicKey(raw []byte, kemId uint16) (hpke.KEMPublicKey, error) {
+	// NOTE: Stand-in values for KDF/AEAD algorithms are ignored.
+	hpkeSuite, err := echAssembleHpkeCipherSuite(kemId, dummyKdfId, dummyAeadId)
+	if err != nil {
+		return nil, err
+	}
+	return hpkeSuite.KEM.Deserialize(raw)
+}
+
+// echUnmarshalHpkeSecretKey parses a serialized secret key for the KEM algorithm
+// identified by `kemId`.
+func echUnmarshalHpkeSecretKey(raw []byte, kemId uint16) (hpke.KEMPrivateKey, error) {
+	// NOTE: Stand-in values for KDF/AEAD algorithms are ignored.
+	hpkeSuite, err := echAssembleHpkeCipherSuite(kemId, dummyKdfId, dummyAeadId)
+	if err != nil {
+		return nil, err
+	}
+	return hpkeSuite.KEM.DeserializePrivate(raw)
+}
+
+// echCreateHpkeKdf returns an HPKE KDF scheme.
+func echCreateHpkeKdf(kdfId uint16) (hpke.KDFScheme, error) {
+	// NOTE: Stand-in values for KEM/AEAD algorithms are ignored.
+	hpkeSuite, err := echAssembleHpkeCipherSuite(dummyKemId, kdfId, dummyAeadId)
+	if err != nil {
+		return nil, err
+	}
+	return hpkeSuite.KDF, nil
+}
+
+func echIsCipherSuiteSupported(suite echCipherSuite) bool {
+	// NOTE: Stand-in values for KEM algorithm is ignored.
+	_, err := echAssembleHpkeCipherSuite(dummyKemId, suite.kdfId, suite.aeadId)
+	return err == nil
+}
+
+func echIsKemSupported(kemId uint16) bool {
+	// NOTE: Stand-in values for KDF/AEAD algorithms are ignored.
+	_, err := echAssembleHpkeCipherSuite(kemId, dummyKdfId, dummyAeadId)
```

Comment:
Hmm, is this right? What if `kemId` is `DHKEM(X25519, SHA-256)` but HPKE only supports `DHKEM(X25519, SHA-256), KDF(SHA-512), AEAD(ChaCha20Poly1305)`? (In other words, HPKE only tells you if the tuple of (KEM, KDF, AEAD) is supported via `AssembleCipherSuite`, not whether a specific KEM is supported for *some* KDF/AEAD pairs.)

Reply:
Oof, you're right. The presumption is that if the HPKE library supports any cipher suite composed of algorithms it implements. In particular, support for (kem1, kdf1, aead1) implies support for (kem2, kdf1, aead1) for any kem1, kem2, kdf1, and aead1 it implements. Looking at github.com/cisco/go-hpke, this appears to be the case today. Perhaps we should add a func init() to the package that panics if this isn't true.

In an earlier version of this PR, I had explicitly defined code points for the HPKE algorithms we intend to support. It seemed superfluous at the time, but it might make sense to bring it back given this external dependency.

Reply:
I think the way to handle this is to do nothing. The only way this would break is if, after updating "github.com/cisco-go-hpke", the MIT cipher suite was no longer supported. This condition is already checked by the unit tests, which attempt to load two different ECHConfigs, one of which is the MTI cipher suite. I will add a note under the dependency.

Reply:
Added a new file, `hpke.go`, which defines the code points we need and ensure that the MTI cipher suite is supported. The unit tests failing will indicate that the upstream HPKE implementation is not suitable for ECH.